### ISSUE TITLE
Explicit simplification in `TrotterProduct` of neighboring identical operators

### ISSUE THREAD

**Issue by Qottmann** [PennyLaneAI/pennylane#6603]
### Feature details

I am getting this:
```
>>> qml.TrotterProduct(X(0) + Y(0), 1., n=3, order=2).decomposition()
[Exp(0.16666666666666666j PauliX),
 Exp(0.16666666666666666j PauliY),
 Exp(0.16666666666666666j PauliY),
 Exp(0.16666666666666666j PauliX),
 Exp(0.16666666666666666j PauliX),
 Exp(0.16666666666666666j PauliY),
 Exp(0.16666666666666666j PauliY),
 Exp(0.16666666666666666j PauliX),
 Exp(0.16666666666666666j PauliX),
 Exp(0.16666666666666666j PauliY),
 Exp(0.16666666666666666j PauliY),
 Exp(0.16666666666666666j PauliX)]
```

whereas I would want to get is
```
[Exp(0.16666666666666666j PauliX),
 Exp(2 * 0.16666666666666666j PauliY),
 Exp(2 * 0.16666666666666666j PauliX),
 Exp(2 * 0.16666666666666666j PauliY),
 Exp(2 * 0.16666666666666666j PauliX),
 Exp(2 * 0.16666666666666666j PauliY),
 Exp(0.16666666666666666j PauliX)]
```

Ideally in both the decomposition and resource estimates for the operator

### How important would you say this feature is?

1: Not important. Would be nice to have.

**Comment by ShreyPatel1311** [comment]
Can I give it a try if it's still open?!

**Comment by CatalinaAlbornoz** [comment]
Hi @ShreyPatel1311, thanks for your interest in this!
Our team is reviewing the request and will get back to you soon.
Please wait for our go-ahead before starting on a fix, as the team may have specific guidance on if or how to tackle this.

**Comment by Jaybsoni** [comment]
Hi @ShreyPatel1311, thanks again for the interest and apologies for the delay! 

You are good to go with starting a fix! I will be one of the developers reviewing your PR. Here are some guidelines for you to keep in mind as you begin the implementation.

The main method you will be updating is the `.compute_decomposition()` function of the `TrotterProduct` class. By extension this will require modifications to the private `_recursive_expression()` function. You should NOT update the private `_trotter_product_decomposition()` or `_trotter_product_decomposition_resources()` functions as part of this bug fix.

Feel free to open up a PR and share the link here. Happy to answer any questions you have and continue the conversation there. Thanks again for your interest in contributing to PennyLane 😄!

**Comment by ShreyPatel1311** [comment]
Hi @Jaybsoni, thanks for assigning this task. Just to make it clear on the issue which I will be working on involves to changes to compute_decomposition() and _recursive_expression() functions.  I think TrotterProduct class is under here("pennylane/templates/subroutines/time_evolution/trotter.py"), right?! 

I have another while I was reviewing this code. What does this apply function do inside this compute_decomposition()?!

**Comment by Jaybsoni** [comment]
> Just to make it clear on the issue which I will be working on involves to changes to compute_decomposition() and _recursive_expression() functions. I think TrotterProduct class is under here("pennylane/templates/subroutines/time_evolution/trotter.py"), right?!

Yes exactly, you will be modifying the code in `pennylane/templates/subroutines/time_evolution/trotter.py` and modifying/adding additional tests in `tests/templates/subroutines/time_evolution/test_trotter.py`

> What does this apply function do inside this compute_decomposition()?

This has to do with something called "queuing" which is the system PennyLane uses to register operator classes that are instantiated inside a qnode into the queue to be executed on the device. Here are some links to help you learn: 
- [Overview of queuing ](https://docs.pennylane.ai/en/stable/code/qml_queuing.html)
- [Specifics of `qp.apply()`](https://docs.pennylane.ai/en/stable/code/api/pennylane.queuing.apply.html)

This is used inside compute decomposition because we require that the list of operators returned by this function should also be queued.

Thanks,

**Comment by ShreyPatel1311** [comment]
Hi @Jaybsoni, so I have completed the task and it is giving me output somewhat like this "
Decomposition:
1 Evolution(0.16666666666666666j PauliX)
2 Evolution(0.3333333333333333j PauliY)
3 Evolution(0.3333333333333333j PauliX)
4 Evolution(0.3333333333333333j PauliY)
5 Evolution(0.3333333333333333j PauliX)
6 Evolution(0.3333333333333333j PauliY)
7 Evolution(0.16666666666666666j PauliX)"

then, I asked codeX to test this particular function which I have created on multiple test cases. It tested on the function present in class TestDecomposition and class TestResources, where it failed since both of these classes where expecting the older version of the output.  Here I have just merged the X and Y  neighboring gates with same angles. I hope that was the kind of feature which you were expecting right?! I haven't made changes to any other functions, which wasn't mentioned in your previous dialogue. So should I make a PR now or I can make to functions in TestDecomposition and TestResources classes and then make a PR?!

**Comment by Jaybsoni** [comment]
Hi @ShreyPatel1311, 

That's awesome work! Yes, now is the perfect time to make a PR to Pennylane. Please tag me as a reviewer on the PR and also share the link here! We do have a formal code review process before anything gets merged into PL, but I can help guide you through that. 

Let me know when you have the PR ready! 
Thanks,

**Comment by ShreyPatel1311** [comment]
Hi @Jaybsoni,

I have submitted the PR for a review. you can find it right here "(https://github.com/PennyLaneAI/pennylane/pull/9262)". Since, this is my first PR I just want to give you the process which I did

-> created a new branch and changes were commit here. Pushed those changes to new branch.
-> updated changlog-dev.md about the update. Did pylint test and made fixes accordingly
-> I didn't what kind of test you are expecting, so I modified the test which I used to get this result and added that to test_trotter.py file
-> Then while submitting review I commented and tagged you as reviewer. I hope that's what I was supposed to do?!

If any more specific changes or updates which you might expect please let me know. I will work on them ASAP!

**Comment by CatalinaAlbornoz** [comment]
Thanks for detailing your process @ShreyPatel1311 ! This is very helpful.

Jay will review and reach out to you directly on the PR. We sometimes take a few weeks to finalize the review but please feel free to tag us on the PR if you haven't heard from us in over 2 weeks.
